src/frame/io.rs

```diff
@@ -93,14 +93,15 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Stream for Io<T> {
             match self.decode() {
                 Ok(Some(f)) => return Poll::Ready(Some(Ok(f))),
                 Ok(None) => {
-                    if self.buffer.remaining_mut() < BLOCKSIZE {
+                    if self.buffer.capacity() - self.buffer.len() < BLOCKSIZE {
                         let n = self.buffer.len();
                         self.buffer.resize(n + BLOCKSIZE, 0);
                         unsafe { self.buffer.set_len(n) }
                     }
                     unsafe {
```

Comment:
Why do we have unsafe here? I'd take a performance hit any day over unsafe code.

Reply:
We need `unsafe` here because of `bytes_mut()` and `advance_mut()`. What we want is reading data into the buffer into the range [len(), capacity()). I do not know an API in `bytes` which would allow us to do this without `unsafe`.